### PR TITLE
ReDo

### LINKED ISSUE #140: ReDO

### PR DESCRIPTION
**Pull request by steventango**

### CODE REVIEW

**Comment by steventango** [comment]
# Code Review

## Overview

Adds **DQN_ReDo**, an implementation of *Recycling Dormant Neurons* (Sokar et al., 2023): every `redo_freq` updates, neurons whose normalized mean activation falls below `redo_threshold` have their incoming weights re-initialized, biases zeroed, outgoing weights zeroed, optionally LayerNorm scale/offset reset, and Adam moments cleared.

Supporting changes:
- `ForagerNet`: new `preactivation_layer_norm` flag (default `True` = unchanged) + surfaces per-layer MLP activations via `accumulatingSequence`.
- `utils/hk.accumulatingSequence`: now also tracks non-`hk.Module` callables (e.g. `jax.nn.relu`) and de-duplicates names with a counter (`relu`, `relu_1`, …) — additive, backward-compatible.
- `NNAgent.AgentState.optim` annotation corrected from \`Dict[str, optax.OptState]\` → \`optax.OptState\` (was wrong; runtime is the optax chain tuple).
- 4 sweep/eval JSONs for \`DQN_ReDo\` and \`DQN_ReDo_PreActLN\` on \`ForagaxSquareWaveTwoBiome-v11\`.
- \`continual-foragax\` 0.51.0 → 0.54.1.

## Correctness

**Bug — ReDo will fire \`update_freq\` times in a row.**
\`src/algorithms/nn/DQN_ReDo.py:174-179\`:
\`\`\`python
do_redo = (
    (state.updates > 0)
    & (state.updates % state.hypers.redo_freq == 0)
    & self.buffer.can_sample(state.buffer_state)
)
\`\`\`
\`_maybe_update\` runs every env step, but \`state.updates\` only increments when \`state.steps % update_freq == 0\`. So once \`updates\` lands on a multiple of \`redo_freq\`, the predicate stays true for the next \`update_freq − 1\` env steps too. With the configs in this PR (\`redo_freq=10000\`, \`update_freq=4\`), ReDo runs 4× back-to-back each scheduled cycle — wasteful and re-resets the just-reset network with fresh buffer samples. \`DQN_Reset\` avoids this by gating on \`state.steps\` (which increments every call). Either gate on \`steps\`, or trigger \`_redo_step\` from inside the \`do_update\` branch.

**Hardcoded Haiku param paths.** \`phi/~/linear\`, \`phi/~/linear_1\`, \`phi/~/layer_norm\`, \`phi/~/layer_norm_1\`, \`q/q/w\` are baked in. Any rename, an extra MLP layer, or a different head silently mis-targets weights without raising. The constructor already rejects \`layers != 2\` / \`conv != \"None\"\` / non-\`ForagerNet\` — consider also asserting these specific keys exist in \`state.params\` at construction so future drift fails loudly.

**Optimizer state indexing.** \`state.optim[0]\` (line 167) assumes the chain is \`(ScaleByAdamState, ScaleByLearningRate)\`. The constructor blocks SWR, so this is currently safe, but adding e.g. weight decay or LR warmup later would silently corrupt the wrong substate. A small \`assert isinstance(state.optim[0], optax.ScaleByAdamState)\` at agent init would be cheap insurance.

**Activation key dependency.** \`_redo_step\` relies on \`feat.activations[\"relu\"]\` and \`feat.activations[\"relu_1\"]\` — those keys come from the new \`accumulatingSequence\` deduplication scheme. Works for both pre- and post-act LN orderings (both put \`relu\` after the first activation function, \`relu_1\` after the second). Worth a one-line comment explaining the mapping.

## Minor

- \`_sample_new_param\` calls \`init_fn(key, ...)\` for **every** \`\"w\"\` leaf even when its \`incoming_mask\` is all-False (e.g. the q-head). Just wasted compute, not a bug.
- \`_use_ln\` / \`_reset_ln\` are stored on \`self\`, not in \`AgentState\`. Fine for the current \`@checkpointable\` keys but means they aren't restored across resumes — if anyone toggles \`redo_reset_layernorm\` mid-run, behavior tracks the new config rather than the old, which is probably what you want anyway.
- JSON style: \`\\t\` mixed with 4-space indent in the new configs (\`			\"observation_type\": \"color\"\`). Matches the existing sibling \`DQN.json\` so this is consistent project drift, not new.
- The PR body is empty — a one-line \"implements Sokar et al. 2023\" plus the supported architecture would help future-you.

## Performance

- \`_redo_step\` is \`jit\`-compiled inside \`lax.cond\`; both branches trace once — fine.
- The mask trees are full pytree-shape booleans broadcast across param shapes. For a 64-wide 2-layer MLP this is trivial; for larger reps it'd be worth thinking about, but irrelevant here.

## Tests / Coverage

No tests. The repo has minimal test coverage for siblings (\`DQN_Reset\`, \`DQN_L2_Init\`, etc.) so this matches convention, but a single smoke test that constructs the agent, forces \`_redo_step\` once, and asserts the dormant neurons' rows changed would catch path-string regressions cheaply.

## Security

No external I/O, no user-controlled inputs. N/A.

## Summary

Implementation is faithful to the ReDo paper and integrates cleanly with the existing DQN scaffold. The main thing worth fixing before merge is the **fire-\`update_freq\`-times** scheduling bug; everything else is brittleness/style that you can punt on for a research artifact.
